core/src/test/scala/com/mesosphere/usi/core/scalaapi/SchedulerAdapterTest.scala

```diff
@@ -0,0 +1,315 @@
+import com.mesosphere.usi.core.models._
+import com.mesosphere.utils.AkkaUnitTest
+import org.scalatest._
+
+import scala.concurrent.Promise
+import scala.util.{Failure, Success}
+
+class SchedulerAdapterTest extends AkkaUnitTest with Inside {
+
+  def createMockedScheduler: (
+      Flow[Scheduler.SpecInput, Scheduler.StateOutput, NotUsed],
+      SinkQueueWithCancel[Scheduler.SpecInput],
+      SourceQueueWithComplete[Scheduler.StateOutput]) = {
+
+    val (outerSinkQueue, outerSink) = Sink.queue[Scheduler.SpecInput]().preMaterialize()
+    val (outerSourceQueue, outerSource) =
+      Source.queue[Scheduler.StateOutput](16, OverflowStrategy.fail).preMaterialize()
+
+    (Flow.fromSinkAndSource(outerSink, outerSource), outerSinkQueue, outerSourceQueue)
+  }
+
+  class SchedulerFixture {
+    val podSpec = PodSpec(
+      PodId("podId"),
+      Goal.Running,
+      RunSpec(
+        resourceRequirements = List(ScalarRequirement.cpus(1), ScalarRequirement.memory(256)),
+        shellCommand = "sleep 3600")
+    )
+    val specEvent = PodSpecUpdated(podSpec.id, None)
+    val stateEvent = PodStatusUpdated(podSpec.id, None)
+
+    val (scheduler, specInputQueue, stateOutputQueue) = createMockedScheduler
+    val adapter = new SchedulerAdapter(scheduler)
+  }
+
+  "SchedulerAdapter" when {
+    "materialized as Source and Sink" should {
+      "Propagate a snapshot to the underlying scheduler" in new SchedulerFixture {
+        adapter.asSourceAndSink(SpecsSnapshot(List(podSpec), Nil))
+        specInputQueue.pull().futureValue.get._1.podSpecs.head shouldEqual podSpec
+      }
+
+      "If snapshot is not provided push an empty snapshot" in new SchedulerFixture {
+        adapter.asSourceAndSink()
+        specInputQueue.pull().futureValue.get._1 shouldEqual SpecsSnapshot.empty
+      }
+
+      "Complete the snapshot promise once it's available" in new SchedulerFixture {
+        val (snapshotF, _, _) = adapter.asSourceAndSink()
+        specInputQueue.pull().futureValue
+
+        val stateSnapshot = StateSnapshot(List(PodStatus(podSpec.id, Map.empty)), Nil, Nil, Nil)
+
+        stateOutputQueue.offer(stateSnapshot -> Source.maybe)
+
+        snapshotF.futureValue shouldEqual stateSnapshot
+      }
+
+      "Push the SpecUpdatedEvents downstream" in new SchedulerFixture {
+        val (_, _, sink) = adapter.asSourceAndSink()
+
+        Source.repeat(specEvent).runWith(sink)
+
+        specInputQueue.pull().futureValue.get._2.runWith(Sink.head).futureValue shouldEqual specEvent
+      }
+
+      "Push the events from scheduler to the provided Source" in new SchedulerFixture {
+        val (_, source, _) = adapter.asSourceAndSink()
+
+        stateOutputQueue.offer(StateSnapshot.empty -> Source.repeat(stateEvent))
+
+        source.runWith(Sink.head).futureValue shouldEqual stateEvent
+      }
+
+      "Complete both sink and source when the event stream coming to scheduler is cancelled by the scheduler" in new SchedulerFixture {
```

Comment:
```suggestion
      "Complete both sink and source when the event stream coming to scheduler is canceled by the scheduler" in new SchedulerFixture {
```

Reply:
This is one of those dumb things where Americans spell it one way, and European spell it another way

Reply:
Well, I'd be consistent with the `Sink.cancelled` method in akka.